### PR TITLE
feat(LH-72231): SEC Resource

### PR DESCRIPTION
**Pull request by weilueluo**
https://jira-eng-rtp3.cisco.com/jira/browse/LH-72231

### Description
- Read SEC
- Update SEC (empty implementation)
- Delete SEC
- Create SEC
	- Create the SEC device in CDO and generate the CDO bootstrap data and SEC bootstrap data. This is only part of the onboarding process.
- Refactored state machine's state field from `string` to concrete `state.Type`

### CODE REVIEW

File: provider/internal/connector/sec/operation.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	planData.CdoBootstrapData = types.StringValue(createOutp.CdoBoostrapData)
+	planData.Name = types.StringValue(createOutp.Name)
+
+	return nil
+}
+
+func Update(ctx context.Context, resource *Resource, planData *ResourceModel, stateData *ResourceModel) error {
+
+	// do update
+	// e.g. updateOutp, err := resource.client.UpdateExample(ctx, ...)
```
Comment (siddhuwarrier) [inline provider/internal/connector/sec/operation.go:49]: Please delete this comment

File: provider/internal/connector/sec/resource.go [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+}
+
+func (r *Resource) Metadata(ctx context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_sec"
+}
+
+func (r *Resource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		// This description is used by the documentation generator and the language server.
+		MarkdownDescription: "TODO",
```
Comment (siddhuwarrier) [inline provider/internal/connector/sec/resource.go:39]: ??????

File: provider/internal/connector/sec/resource.go [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+			},
+			"name": schema.StringAttribute{
+				MarkdownDescription: "TODO",
+				Computed:            true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+			"cdo_bootstrap_data": schema.StringAttribute{
+				MarkdownDescription: "TODO",
```
Comment (siddhuwarrier) [inline provider/internal/connector/sec/resource.go:57]: ??????????????????

File: provider/internal/connector/sec/resource.go [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+				Computed:            true,
+				Sensitive:           true, // bootstrap data contains user api token
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+			"sec_bootstrap_data": schema.StringAttribute{
+				MarkdownDescription: "TODO",
```
Comment (siddhuwarrier) [inline provider/internal/connector/sec/resource.go:65]: ???????

**Review by siddhuwarrier — CHANGES_REQUESTED** [review]

File: client/connector/sec/create.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+			EarlyExitOnError(true).
+			Timeout(5*time.Minute).
+			Retries(-1).
+			Delay(time.Second).
+			Build(),
+	)
+	if err != nil {
+		return nil, err
+	}
+	// 3. get sec bootstrap data
```
[the remark this answers is not in the document]
Comment (weilueluo) [reply client/connector/sec/create.go:52]: done! 👍

File: provider/internal/cdfmc/resource_test.go [hunk 6]
```diff
@@ -20,7 +20,7 @@ resource "cdo_cdfmc" "test" {
 var resourceConfig = acctest.MustParseTemplate(resourceTemplate, resourceModel)
 
 func TestAccCdFmcResource(t *testing.T) {
-	t.Skip("we cant delete a fmc so this test cannot be run, or we should find a way to spin up new environment, either seems uneasy, skipping for now.")
+	t.Skip("we cant delete a fmc so this test cannot be run, or we should find a way to spin up new environment, neither seems uneasy, skipping for now.")
```
[the remark this answers is not in the document]
Comment (weilueluo) [reply provider/internal/cdfmc/resource_test.go:23]: done! 👍

File: provider/internal/connector/sec/operation.go (hunk 1 above)
Comment (weilueluo) [reply provider/internal/connector/sec/operation.go:49]: done! 👍

File: provider/internal/connector/sec/operation.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+	_, err := resource.client.UpdateSec(ctx, sec.NewUpdateInputBuilder().Build())
+	if err != nil {
+		return err
+	}
+	// map response to terraform types
+	// stateData.ID = types.StringValue(updateOutp.Uid)
```
[the remark this answers is not in the document]
Comment (weilueluo) [reply provider/internal/connector/sec/operation.go:55]: done! 👍

File: provider/internal/connector/sec/resource.go (hunk 4 above)
Comment (weilueluo) [reply provider/internal/connector/sec/resource.go]: done 👍

File: provider/internal/connector/sec/resource.go (hunk 3 above)
Comment (weilueluo) [reply provider/internal/connector/sec/resource.go]: done 👍

File: provider/internal/connector/sec/resource.go [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				MarkdownDescription: "TODO",
+				Computed:            true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+			"name": schema.StringAttribute{
+				MarkdownDescription: "TODO",
```
[the remark this answers is not in the document]
Comment (weilueluo) [reply provider/internal/connector/sec/resource.go]: done 👍

File: client/connector/sec/create_test.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+	"github.com/CiscoDevnet/terraform-provider-cdo/go-client/examples"
+	internalHttp "github.com/CiscoDevnet/terraform-provider-cdo/go-client/internal/http"
+	"github.com/CiscoDevnet/terraform-provider-cdo/go-client/internal/url"
+	"github.com/jarcoal/httpmock"
+	"net/http"
+	"testing"
+	"time"
+)
+
+func TestSECCreate(t *testing.T) {
```
Comment (hazital) [inline client/connector/sec/create_test.go:14]: pls add tests for the delete, read, and update as well

**Comment by weilueluo** [comment]
superseded by https://github.com/CiscoDevNet/terraform-provider-cdo/pull/104

**Pull request closed without merging**